### PR TITLE
[HWKMETRICS-172] Add checks for empty or missing payloads for various…

### PR DESCRIPTION
**Pull request by stefannegrea**
… types to allow users to omit them from the request. Also, added tests for various scenarios.

### CODE REVIEW

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java [hunk 1]
```diff
@@ -101,20 +107,34 @@ public void findMetrics(
     @ApiOperation(value = "Add data for multiple metrics in a single call.")
     @ApiResponses(value = {
             @ApiResponse(code = 200, message = "Adding data succeeded."),
+            @ApiResponse(code = 400, message = "Missing or invalid payload.", response = ApiError.class),
```
Comment (tsegismont) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java]: I couldn't find the code checking for an empty payload. Did I miss it?

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/request/MixedMetricsRequest.java [hunk 2]
```diff
@@ -42,15 +42,27 @@
...
     }
 
     public List<Counter> getCounters() {
-        return ImmutableList.copyOf(counters);
+        if (counters != null) {
+            return ImmutableList.copyOf(counters);
+        }
+
+        return null;
     }
```
Comment (tsegismont) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/request/MixedMetricsRequest.java]: Wondering if we shouldn't simply drop the `ImmutableList.copyOf` and thus the new null check, as this a mutable object on purpose (representation of the mixed metrics entity) and not used anywhere else than in the handler code.

File: rest-tests/src/test/groovy/org/hawkular/metrics/rest/MetricsITest.groovy [hunk 3]
```diff
@@ -150,4 +150,189 @@ class MetricsITest extends RESTTest {
...
+        path: "metrics/data",
+        headers: [(tenantHeaderName): tenantId],
+        body: [
+            gauges: [
+            ],
+            counters: [
+            ]
+        ]
+    )
+    assertEquals(200, response.status)
```
Comment (tsegismont) [inline rest-tests/src/test/groovy/org/hawkular/metrics/rest/MetricsITest.groovy]: Shouldn't the status code be 400? Gauge handler replies with 400 when the list of metrics is empty.

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java [hunk 4]
```diff
@@ -101,20 +107,34 @@ public void findMetrics(
...
             @Suspended final AsyncResponse asyncResponse,
             @ApiParam(value = "List of metrics", required = true) MixedMetricsRequest metricsRequest) {
 
-        Observable<Void> gaugeInserts = metricsService.addGaugeData(requestToGauges(tenantId,
-                metricsRequest.getGaugeMetrics()).subscribeOn(Schedulers.computation()));
-        Observable<Void> counterInserts = metricsService.addCounterData(requestToCounters(tenantId,
-                metricsRequest.getCounters()).subscribeOn(Schedulers.computation()));
-        Observable<Void> availabilityInserts = metricsService.addAvailabilityData(requestToAvailabilities(tenantId,
-                metricsRequest.getAvailabilityMetrics()).subscribeOn(Schedulers.computation()));
+        Collection<Observable<Void>> observables = new ArrayList<>();
```
Comment (burmanm) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java]: This isn't very reactive design and actually a bit more difficult to follow what's happening. Having empty observables is not an issue, they are just not processed (nothing to react to).

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java (hunk 1 above)
Comment (tsegismont) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java]: Now I understand what's going on. In `MetricsITest#addMixedDataEmptyRequestPayload`, you set the body to `[]`. But this is not an empty object, this is an empty array. So Jackson fails to deserialize from an array, throws an exception, converted by RestEasy into a 400 reply. 

What you need to do to test the empty object case is `body: new HashMap<>()`.

Also there's the null test case which is missing. And if you add it, you'll get a 500 reply, because of an NPE thrown in this method. You need to check if `metricsRequest` is null and return `emptyPayload` in this case.

**Comment by tsegismont** [comment]
Stefan are your going to update all `@POST` endpoints in this PR? Or other PRs?

**Comment by stefannegrea** [comment]
@tsegismont, I will create another Jira/PR to update the rest of the @POST methods. I want to keep this PR only for the newly updated method.

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java (hunk 1 above)
Comment (stefannegrea) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java]: I updated the code and test cases to check for empty payload early and return 400 if no content.

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java (hunk 4 above)
Comment (stefannegrea) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/MetricHandler.java]: Agree, I updated the code to preempt executing that code when there is no content.

File: rest-tests/src/test/groovy/org/hawkular/metrics/rest/MetricsITest.groovy (hunk 3 above)
Comment (stefannegrea) [inline rest-tests/src/test/groovy/org/hawkular/metrics/rest/MetricsITest.groovy]: You are right, we should treat empty content consistently and return 400 even though the named arrays are there but empty. Updated the code to reflect this.

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/request/MixedMetricsRequest.java (hunk 2 above)
Comment (stefannegrea) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/request/MixedMetricsRequest.java]: I dropped the immutable portion out of the request object and closer to the metrics service calls. That should help checking for null and empty content without incurring the copy penalty.

**Comment by tsegismont** [comment]
Ha! The build failed again because of the `CountersITest` intermittent failure:
```
Failed tests: 
  CountersITest.findRate:288 expected:<200> but was:<204>
```

I'll merge the PR anyway as the other tests are passing.
